### DISCUSSION TITLE
Make a method call after order submitted

### DISCUSSION THREAD

**Discussion by VladimirBV** [SalesforceCommerceCloud/sfcc-ci#427] [Team Posts]
Hi,
I need to write a prototype with my own cartridge, I need to call a method dw.order.Order.**getOrderExportXML**(..) after order submitted(payed) and send a file to FTP. How can I do this?

**Comment by zlatinz** [comment]
Hi @VladimirBV,
There are no DB triggers or alike functionality in Commerce Cloud.
Hence the only viable option to solve this is to create a job, that looks for orders that match your criteria (paid order) and do the operations you have mentioned in the job. 
You cannot do these in storefront context, as you mentioned that you need to create a file (storefront quota is 0) and send it to FTP server (which can be longer than acceptable for storefront context)

**Comment by VladimirBV** [comment]
yes, not a file but a string with xml, so maybe to send to mule api web service

**Comment by pravngaur** [comment]
Hi, if file creation is not required and if you are on SFRA then you could
create the logic of calling the api and plugin it in as a middleware on the
onComplete event.

Also, please do not call the muel api synchronously as it will impact
storefront performance rather call asynchronously using a promise(or
timeout).

**Comment by zlatinz** [comment]
I would not call this type of a backend integration from client-side. Promises are not supported server-side. Everything you do on server-side controller, even if it is event-based is run in synchronous manner.

**Comment by hnestmann** [comment]
Why didn't anyone suggest the commerce cloud default approach? Have a job run every minute or so which writes order files to disk and puts its on the ftp? 
See this for a couple of examples: https://github.com/SalesforceCommerceCloud/job-components
This allows scalable order intake and steady delivery to the OMS/Warehouse

**Comment by pravngaur** [comment]
Hello Zlatin,

Promises are supported server side, it is being used in b2c connectors,
that's how i now it works; let me know if that's not the case.

Holger, not always FTP offline approach works in requirements where real
time synchup is required, what he is explicitly looking for is and
mentioned is real time integration.

Vladimir, now you have all the options, see which one works for you

**Comment by zlatinz** [comment]
Promises are emulated server-side using https://github.com/SalesforceCommerceCloud/marketing-cloud-connector/blob/205f76731bb78c4589791648e07877715a57486e/cartridges/synchronous-promise.js in MC connector, but there is no native platform support for them. They are also run in a synchronous manner, as is everything that is run server-side.
@hnestmann, if you look at my comment (first comment under this question) https://github.com/orgs/SalesforceCommerceCloud/teams/community/discussions/77/comments/1 I am suggesting this to be solved by job, as jobs are the only way to have asynchronous-like communication, and they allow for longer timeouts on communication.
Real-time is one of the most abused terms when it comes to e-commerce sites. No any process happens immediately, there is always sequence of events. FTP by definition requires file transfer, and there is no way the FTP server to notify "subscribers" (as there are no "subscribers"), so the velocity and delays of data updates really depend on how often information from the FTP server is pulled about the files stored there.

**Comment by VladimirBV** [comment]
>Why didn't anyone suggest the commerce cloud default approach?
by law order should be sent immediately, without any delay..

**Comment by zlatinz** [comment]
@VladimirBV typical design-pattern is to do synchronous integration (using API) after order submitted (paid) and have a fallback job, that re-processes orders that has failed this synchronous integration.
Once again if you have to create files from storefront context - this is not possible and is not a scalable solution design, and you need to do this from a job. Running the job on frequent intervals (like every couple of minutes or so) should suffice. Network delays are inevitable, I do not see why a minute delay (for technical reasons) would be a problem.
Hope it makes sense.
